src/solvers/elliptic/linearSolver/PCG_EIGEN.cpp

```diff
@@ -0,0 +1,339 @@
+  occa::memory &o_p  = elliptic->o_p;
+  occa::memory &o_z = (!options.compareArgs("PRECONDITIONER", "NONE")) ? elliptic->o_z : o_r;
+  occa::memory &o_Ap = elliptic->o_Ap;
+  occa::memory &o_weight = elliptic->o_invDegree;
+  platform->linAlg->fill(elliptic->Nfields * elliptic->fieldOffset, 0.0, o_p);
+
+  if(platform->comm.mpiRank == 0 && verbose) {
+    if(flexible) 
+      printf("PFCG ");	  
+    else
+      printf("PCG ");	  
+    printf("%s: initial res norm %.15e WE NEED TO GET TO %e \n", elliptic->name.c_str(), rdotr, tol);
+  }
+
+  int iter = 0;
+  do {
+    iter++;
+    const dfloat rdotz2 = rdotz1;
+    if(!options.compareArgs("PRECONDITIONER", "NONE")) {
+      ellipticPreconditioner(elliptic, o_r, o_z);
+
+      rdotz1 = platform->linAlg->weightedInnerProdMany(
+        mesh->Nlocal,
+        elliptic->Nfields,
+        elliptic->fieldOffset,
+        o_weight,
+        o_r,
+        o_z,
+        platform->comm.mpiComm);
+    } else {
+      rdotz1 = rdotr; 
+    }
+
+#ifdef DEBUG
+    printf("norm rdotz1: %.15e\n", rdotz1);
+#endif
+
+    dfloat beta = 0;
+    if(iter > 1) {
+      beta = rdotz1/rdotz2;
+      if(flexible) {
+        const dfloat zdotAp = platform->linAlg->weightedInnerProdMany(
+          mesh->Nlocal,
+          elliptic->Nfields,
+          elliptic->fieldOffset,
+          o_weight,
+          o_z,
+          o_Ap,
+          platform->comm.mpiComm);
+        beta = -alpha * zdotAp/rdotz2;
+#ifdef DEBUG
+        printf("norm zdotAp: %.15e\n", zdotAp);
+#endif
+      }
+    }
+
+#ifdef DEBUG
+        printf("beta: %.15e\n", beta);
+#endif
+
+
+    platform->linAlg->axpbyMany(
+      mesh->Nlocal,
+      elliptic->Nfields,
+      elliptic->fieldOffset,
+      1.0,
+      o_z,
+      beta,
+      o_p);
+
+    const dfloat pAp0 = pAp;
+    ellipticOperator(elliptic, o_p, o_Ap, dfloatString);
+    pAp = platform->linAlg->weightedInnerProdMany(
+      mesh->Nlocal,
+      elliptic->Nfields,
+      elliptic->fieldOffset,
+      o_weight,
+      o_p,
+      o_Ap,
+      platform->comm.mpiComm);
+    alpha = rdotz1 / (pAp + 1e-300);
+
+#ifdef DEBUG
+    printf("alpha: %.15e\n", alpha);
+    printf("norm pAp: %.15e\n", pAp);
+#endif
+
+    //  x <= x + alpha*p
+    //  r <= r - alpha*A*p
+    //  dot(r,r)
+    rdotr = sqrt(update(elliptic, o_p, o_Ap, alpha, o_x, o_r) * elliptic->resNormFactor);
+#ifdef DEBUG
+    printf("rdotr: %.15e\n", rdotr);
+#endif
+
+    if (iter==1) {
+      diagt[iter-1] = pAp / rdotz1;
+    } else {
+      diagt[iter-1] = (beta*beta*pAp0 + pAp) / (rdotz1 + 1e-300);
+      upper[iter-1] = -beta*pAp0 / sqrt(rdotz2*rdotz1);
+    }
+    if (platform->comm.mpiRank == 0)
+      nrsCheck(std::isnan(rdotr), MPI_COMM_SELF, EXIT_FAILURE,
+               "%s\n", "Detected invalid resiual norm while running linear solver!");
+
+    if (verbose && (platform->comm.mpiRank == 0))
+      printf("it %d r norm %.15e\n", iter, rdotr);
+  }
+  while (rdotr > tol && iter < MAXIT);
+
+  if (iter>=3) {
```

Comment:
In the new version, I only save the dmin/dmax if iter>=3 and put a integer isEigenReady there.
If the chebyshev is about to be called while isEigenReady is still 0 (unset), it will throw error message.